### PR TITLE
grpcproxy: split out / tighten up leadership detection

### PR DESCRIPTION
**Pull request by heyitsanthony**
Move leadership detection out of watch code since it will be shared with
lease code. Also stricter: assumes leader does not exist unless watch can
be successfully created.

### CODE REVIEW

**Comment by fanminshi** [comment]
lgtm

File: proxy/grpcproxy/leader.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+	close(l.leaderc)
+	go l.recvLoop()
+	return l
+}
+
+func (l *leader) recvLoop() {
+	defer close(l.donec)
+
+	limiter := rate.NewLimiter(rate.Limit(retryPerSecond), retryPerSecond)
+	rev := int64((uint64(1) << 63) - 2)
```
Comment (xiang90) [inline proxy/grpcproxy/leader.go]: math.Maxint64?

File: proxy/grpcproxy/watch.go [hunk 2]
```diff
@@ -108,7 +92,15 @@ func (wp *watchProxy) Watch(stream pb.Watch_WatchServer) (err error) {
 	if md, ok := metadata.FromContext(stream.Context()); ok {
 		v := md[rpctypes.MetadataRequireLeaderKey]
 		if len(v) > 0 && v[0] == rpctypes.MetadataHasLeader {
-			leaderc = wp.lostLeaderNotify()
+			leaderc = wp.leader.lostNotify()
```
Comment (xiang90) [inline proxy/grpcproxy/watch.go]: lostLeaderC?

Comment (xiang90) [inline proxy/grpcproxy/watch.go]: this change makes line 99 reads better.

**Comment by xiang90** [comment]
LGTM

**Pull request merged**